src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -533,6 +541,46 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
             location=node.location,
         )
 
+    def _extract_promoted_tuple_type(
+        self,
+        node_type_ls: list,
+        mask_type: ct.FieldType,
+        return_type: None | ct.FieldType | ct.ScalarType,
+    ) -> ct.FieldType:
+        for element in node_type_ls:
+            if isinstance(element, ct.TupleType):
+                return self._extract_promoted_tuple_type(element.types, mask_type, return_type)
+            else:
+                if return_type is None:
+                    return_type = element
+                else:
+                    return_type = type_info.promote(return_type, element)
+                    if isinstance(return_type, ct.ScalarType) or not all(
+                        item in return_type.dims for item in mask_type.dims
+                    ):
+                        return_type = self._promote_to_mask_type(mask_type, return_type)
+        return return_type
+
+    def _construct_tuple_type(
+        self,
+        node_type_ls: list,
+        promoted_field_tuple: ct.FieldType | ct.TupleType,
+        single_tuple: bool,
+    ) -> ct.TupleType:
+        for i, element in enumerate(node_type_ls):
+            if not isinstance(element, ct.TupleType):
+                node_type_ls[i] = promoted_field_tuple
+            else:
+                self._construct_tuple_type(element.types, promoted_field_tuple, single_tuple)
+        return node_type_ls
+
+    def _promote_to_mask_type(
+        self, mask_type: ct.FieldType, input_type: ct.FieldType | ct.ScalarType
+    ) -> ct.FieldType:
+
+        return_dtype = input_type.dtype if isinstance(input_type, ct.FieldType) else input_type
+        return type_info.promote(input_type, ct.FieldType(dims=mask_type.dims, dtype=return_dtype))
```

Comment:
Since `_extract_promoted_tuple_type`, `_construct_tuple_type`, `_promote_to_mask_type` deal with types and none of them makes use of the `self` instance, except to call themselves, I would suggest making them free functions. The precedent for this is the `boolified_type` function in the same file.